src/components/Editor/Tab.js

```diff
@@ -0,0 +1,217 @@
+
+import CloseButton from "../shared/Button/Close";
+
+import type { List } from "immutable";
+import type { SourceRecord } from "../../reducers/sources";
+
+import { getFilename, getFileURL, isPretty } from "../../utils/source";
+import { copyToTheClipboard } from "../../utils/clipboard";
+import classnames from "classnames";
+
+type SourcesList = List<SourceRecord>;
+
+type Props = {
+  tabSources: SourcesList,
+  selectSource: Object => void,
+  selectedSource: SourceRecord,
+  closeTab: string => void,
+  closeTabs: (List<string>) => void,
+  togglePrettyPrint: string => void,
+  showSource: string => void,
+  source: SourceRecord,
+  sourceTabsMetaData: {
+    [key: string]: SourceMetaDataMap
+  }
+};
+
+export default class Tab extends PureComponent<Props, State> {
+  constructor() {
+    super();
+  }
+
+  onTabContextMenu = (event, tab: string) => {
+    event.preventDefault();
+    this.showContextMenu(event, tab);
+  };
+
+  getSourceAnnotation(source) {
+    const sourceId = source.get("id");
+    const sourceMetaData = this.props.sourceTabsMetaData[sourceId];
+
+    if (sourceMetaData && sourceMetaData.isReactComponent) {
+      return <img className="react" />;
+    }
+    if (isPretty(source)) {
+      return <img className="prettyPrint" />;
+    }
+    if (source.get("isBlackBoxed")) {
+      return <img className="blackBox" />;
+    }
+  }
+
+  showContextMenu(e, tab) {
+    const { closeTab, tabSources, showSource, togglePrettyPrint } = this.props;
+
+    const closeTabLabel = L10N.getStr("sourceTabs.closeTab");
+    const closeOtherTabsLabel = L10N.getStr("sourceTabs.closeOtherTabs");
+    const closeTabsToEndLabel = L10N.getStr("sourceTabs.closeTabsToEnd");
+    const closeAllTabsLabel = L10N.getStr("sourceTabs.closeAllTabs");
+    const revealInTreeLabel = L10N.getStr("sourceTabs.revealInTree");
+    const copyLinkLabel = L10N.getStr("copySourceUri2");
+    const prettyPrintLabel = L10N.getStr("sourceTabs.prettyPrint");
+
+    const closeTabKey = L10N.getStr("sourceTabs.closeTab.accesskey");
+    const closeOtherTabsKey = L10N.getStr(
+      "sourceTabs.closeOtherTabs.accesskey"
+    );
+    const closeTabsToEndKey = L10N.getStr(
+      "sourceTabs.closeTabsToEnd.accesskey"
+    );
+    const closeAllTabsKey = L10N.getStr("sourceTabs.closeAllTabs.accesskey");
+    const revealInTreeKey = L10N.getStr("sourceTabs.revealInTree.accesskey");
+    const copyLinkKey = L10N.getStr("copySourceUri2.accesskey");
+    const prettyPrintKey = L10N.getStr("sourceTabs.prettyPrint.accesskey");
+
+    const tabIDs = tabSources.map(t => t.get("id"));
+    const otherTabs = tabSources.filter(t => t !== tab);
+    const tabID = tabIDs.find(t => t == tab);
+    const tabURLs = tabSources.map(t => t.get("url"));
+    const otherTabURLs = otherTabs.map(t => t.get("url"));
+
+    if (!tabID) {
+      return;
+    }
+
+    const isPrettySource = isPretty(tabID);
+
+    const closeTabMenuItem = {
```

Comment:
Feel like these menu items could be moved to a utility file for most of their body, something like

```js
export const tabMenuItems = {
  closeTab: {
    id: "node-menu-close-tab",
    label: L10N.getStr("sourceTabs.closeTab"),
    accesskey: L10N.getStr("sourceTabs.closeTab.accesskey"),
    disabled: false
  },
  ...
};
```

Then this way when you're building up the list here all you're adding is the `click` properties and can spread in the hard coded stuff from elsewhere.

Reply:
yep - lets do this in a separate PR though to keep this smaller